### ISSUE TITLE
[BUG]: The media file name with square brackets was not processed correctly

### ISSUE THREAD

**Issue by ChuckieChen945** [ankidroid/Anki-Android#17456]
### Checked for duplicates?

- [x] This issue is not a duplicate

### Does it also happen in the desktop version?

- [x] This bug does not occur in the latest version of Anki Desktop

### What are the steps to reproduce this bug?

1. Reference some media files in the card template, with some filenames containing square brackets (not necessarily in pairs), as shown below:
    ```
    {{front}}</br>

    <img src="_heart[123].png">
    <img src="_heart[_test.png">
    <img src="_heart.png">
    ```
2. Study a card using this template
3. Media files with filenames containing square brackets fail to display properly

![Image](https://github.com/user-attachments/assets/9e8d9881-fa22-49b1-a4f4-053190102cbe)

### Expected behaviour

Media files with filenames containing square brackets should be handled correctly, as they are in Anki Desktop:

![Image](https://github.com/user-attachments/assets/3a375f10-7438-4121-ba49-d8e7abaacb48)

### Debug info

```text
AnkiDroid Version = 2.20alpha1 (edd1c61e6445ccac2ddf3cabe0d63c6a303f27b7)

Backend Version = 0.1.43-anki24.10rc2 (24.10rc2 898c91284bc97c928adbf1bcc8ca885b068dac5f)

Android Version = 10 (SDK 29)

ProductFlavor = full

Manufacturer = HUAWEI

Model = MXW-AN00

Hardware = mt6873

Webview User Agent = Mozilla/5.0 (Linux; Android 10; MXW-AN00 Build/HONORMXW-AN00; wv) AppleWebKit/537.36 (KHTML, like Gecko) Version/4.0 Chrome/99.0.4844.88 Mobile Safari/537.36

ACRA UUID = 415b3982-d60c-4528-9002-85dc99d4ac32

FSRS = 1.4.3 (Enabled: true)

Crash Reports Enabled = true
```

### Research

- [x] I have checked the [manual](https://ankidroid.org/docs/manual.html) and the [FAQ](https://github.com/ankidroid/Anki-Android/wiki/FAQ) and could not find a solution to my issue
- [x] (Optional) I have confirmed the issue is not resolved in the latest alpha release ([instructions](https://docs.ankidroid.org/manual.html#betaTesting))

**Comment by BrayanDSO** [comment]
Escape the brackets and it should work.

**Comment by ChuckieChen945** [comment]
Thank you for your suggestion. I’ve tried the following code:

```html
{{front}}</br>

<img src="_heart&#91;123&#93;.png">
<img src="_heart&#91;_test.png">
<img src="_heart.png">
```

and also:

```html
{{front}}</br>

<img src="_heart\[123\].png">
<img src="_heart\[_test.png">
<img src="_heart.png">
```

However, escaping the brackets using `&#91;` and `&#93;` or `\` did not resolve the issue. Neither the manual nor the FAQ provides a solution for this specific problem. Did I miss something? If so, could you please clarify further?

It’s worth noting that using square brackets (`[ ]`) in filenames works perfectly fine in Anki Desktop without any need for escaping. This inconsistency between AnkiDroid and Anki Desktop appears to be an Anki Ecosystem Compatibility issue.

I believe it would be ideal to align AnkiDroid's behavior with Anki Desktop to ensure a consistent user experience across platforms. Thank you for considering this, and please let me know if you need any additional information or testing!

**Comment by lukstbit** [comment]
@ChuckieChen945 I just tested this out of curiosity and it works. Did you try closing those _img_ tags: 
`<img src="_heart[123].png" />`
?

**Comment by ChuckieChen945** [comment]
After further investigation, I discovered that media files containing square brackets cannot be synced via AnkiWeb. It appears that Anki does not support special characters, including square brackets, in media filenames.

When you add media to a field—either by using the paperclip icon in the editor or by pasting it into a field—Anki automatically handles special characters. However, referencing media files with special characters directly in templates may lead to issues.

Thank you again for your time and support in addressing this matter! I’m leaving this information here as a reference for others who may encounter similar problems.

**Comment by ShiroiKuma0** [comment]
This issue also affects users whose **AnkiDroid storage directory path** contains square brackets (not just media filenames). For example, a directory like `/storage/emulated/0/〇/[271] 暗記ドロイド/collection.media/` causes a `java.net.URISyntaxException: Illegal character in path at index 29` when playing any media file.

The root cause is `getFileUri()` in `Sound.kt`, which passes the path to `java.net.URI("file", path, null)`. The `URI` constructor treats `[` and `]` as illegal characters.

### Proposed fix

Percent-encode square brackets before constructing the URI:

```kotlin
fun getFileUri(path: String): URI {
    var p = path
    if (File.separatorChar != '/') p = p.replace(File.separatorChar, '/')
    if (!p.startsWith("/")) p = "/$p"
    if (!p.startsWith("//")) p = "//$p"
    // Percent-encode square brackets which are illegal in java.net.URI
    p = p.replace("[", "%5B").replace("]", "%5D")
    return URI("file", p, null)
}
```

This worked in 2.16.5 because the old version used `File`-based paths internally rather than `URI`-based paths, and `java.io.File` handles brackets without issue.

I have built and tested this fix on Android 13 with AnkiDroid 2.24.0alpha12 and it resolves the `URISyntaxException`.
